Add docs for KEP-3329 Retriable and non-retriable Pod failures for Jobs (#35219)
kubernetes/website: content/en/docs/concepts/workloads/pods/disruptions.md

- the type of controller
- the cluster's resource capacity

## Pod disruption conditions {#pod-disruption-conditions}

{{< feature-state for_k8s_version="v1.25" state="alpha" >}}

When you enable the `PodDisruptionsCondition` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
in your cluster, a dedicated Pod `DisruptionTarget` condition is added to indicate

> a disruption. Its `reason` field additionally indicates one of the following

Reviewer: ```suggestion
an imminent disruption. Its `reason` field additionally indicates one of the following
```

Or some alternative way of saying that the disruption didn't happen yet, but it's about to happen.

Also, you might want to explain that the disruption might be interrupted (because of transient failures). The control plane might continue with the disruption, but its not guaranteed. After some time, the disruption target will be cleared if the conditions in the cluster changed.

reasons for the Pod termination:

> - `PreemptionByKubeScheduler` (Pod preempted by kube-scheduler)

Reviewer: ```suggestion
- `PreemptionByKubeScheduler` (Pod preempted by kube-scheduler to accommodate a pod with higher priority)
```

> - `DeletionByTaintManager` (Pod deleted by taint manager due to NoExecute taint)

Reviewer: link to https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions

Author: Added

> - `EvictionByEvictionAPI` (Pod evicted by Eviction API)

Reviewer: link to https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/

Author: Added

> - `DeletionByPodGC` (an orphaned Pod deleted by PodGC)

Reviewer: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection

> You may want to use these pod disruption conditions you create your

Reviewer: ```suggestion
When using a Job, you may want to use these pod disruption conditions you create your
```

- the type of controller
- the cluster's resource capacity

## Pod disruption conditions {#pod-disruption-conditions}

{{< feature-state for_k8s_version="v1.25" state="alpha" >}}

When you enable the `PodDisruptionsCondition` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)

> in your cluster, a dedicated Pod `DisruptionTarget` condition is added to indicate

Reviewer: Worth highlighting this section (the need to enable the feature gate) in a `note`.

Author: Moved into note

an imminent disruption. Its `reason` field additionally indicates one of the following
reasons for the Pod termination:
- `PreemptionByKubeScheduler` (Pod preempted by kube-scheduler to accommodate a pod with higher priority, see more [here](/docs/concepts/scheduling-eviction/pod-priority-preemption/))
- `DeletionByTaintManager` (Pod deleted by taint manager due to NoExecute taint, see more [here](/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions))
- `EvictionByEvictionAPI` (Pod evicted by [Eviction API](/docs/concepts/scheduling-eviction/api-eviction/))
- `DeletionByPodGC` (an orphaned Pod deleted by [PodGC](/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection))

{{< note >}}
A Pod disruption might be interrupted. The control plane might re-attempt to
continue the disruption of the same Pod, but it is not guaranteed. As a result,
the `DisruptionTarget` condition might be added to Pod, but the Pod might not be

> deleted for a long time. In such a situation, after some time, the

Reviewer: ```suggestion
deleted. In such a situation, after some time, the
```

- the type of controller
- the cluster's resource capacity

## Pod disruption conditions {#pod-disruption-conditions}

{{< feature-state for_k8s_version="v1.25" state="alpha" >}}

{{< note >}}
In order to use this behavior, you must enable `PodDisruptionsCondition`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
in your cluster.
{{< /note >}}

When enabled, a dedicated Pod `DisruptionTarget` condition is added to indicate

> an imminent disruption of a Pod. The `reason` field of the condition additionally

Author: Rephrased as: "... that the Pod is about to be deleted due to a disruption"

indicates one of the following reasons for the Pod termination:

> - `PreemptionByKubeScheduler` (Pod preempted by kube-scheduler to accommodate a Pod with higher priority, see more [here](/docs/concepts/scheduling-eviction/pod-priority-preemption/))

Reviewer 2: ```suggestion
- `PreemptionByKubeScheduler`: Pod preempted by kube-scheduler to accommodate a Pod with higher priority. For more information, see [Pod priority preemption](/docs/concepts/scheduling-eviction/pod-priority-preemption/).
```

- `DeletionByTaintManager` (Pod deleted by taint manager due to NoExecute taint, see more [here](/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions))

> - `EvictionByEvictionAPI` (Pod evicted by [Eviction API](/docs/concepts/scheduling-eviction/api-eviction/))

Reviewer 2: ```suggestion
- `EvictionByEvictionAPI`: Pod evicted by [Eviction API](/docs/concepts/scheduling-eviction/api-eviction/).
```

> - `DeletionByPodGC`: an orphaned Pod deleted by [PodGC](/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection).

Author: - the taint manager actually deletes the pod (not kubelet) in this case: https://github.com/kubernetes/kubernetes/blob/132f29769dfecfc808adc58f756be43171054094/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L140
- in PodGC we actually add the disruption condition only if it is orphaned (scheduled to an unavailable Node): https://github.com/kubernetes/kubernetes/blob/132f29769dfecfc808adc58f756be43171054094/pkg/controller/podgc/gc_controller.go#L220

> - `DeletionByPodGC`: an orphaned Pod deleted by [PodGC](/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection).

Reviewer 3: - Can we mention that the taint manager is part of the node lifecycle controller? Most readers won't have heard of the taint manager; I hadn't (and guessed it was part of the kubelet).

- I wouldn't use _orphaned_ to mean “scheduled to an unavailable Node” as the Kubernetes project usually uses the orphan jargon in connection with `.metadata.ownerReferences`.

> - `DeletionByPodGC`: an orphaned Pod deleted by [PodGC](/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection).

Author: @sftim I moved the cleanup to the follow-up PR: https://github.com/kubernetes/website/pull/36007

{{< note >}}
A Pod disruption might be interrupted. The control plane might re-attempt to
continue the disruption of the same Pod, but it is not guaranteed. As a result,
the `DisruptionTarget` condition might be added to Pod, but the Pod might not be
deleted. In such a situation, after some time, the
Pod disruption condition will be cleared.
{{< /note >}}

> When using a Job, you may want to use these Pod disruption conditions you create your

Reviewer 2: ```suggestion
When using a Job, you may want to use these Pod disruption conditions you defined in your
```
